### PR TITLE
New brancher using counting-based search with maxSD

### PR DESCRIPTION
**Pull request by SaGagnon**
Hi Christian,

Here is a new brancher that uses solution densities to branch on <variable, value> pairs with maximum solution densities. I also added it as a branching option in examples/qcp.cpp as a first working example. Using IPL_DOM, I benchmarked our brancher maxSD vs the size and afc strategies on instances currently in qcp.cpp:

[QCP.pdf](https://github.com/Gecode/gecode/files/2334065/QCP.pdf)

As we can see, it offers a small improvement. However, the provided instances were not very challenging, so I added 40 hard instances to qcp.cpp and made a comparison using those:

[QCP_dom_30.pdf](https://github.com/Gecode/gecode/files/2334070/QCP_dom_30.pdf)

Everything compiles both with and without --enable-cbs. I can also make a new pull request against develop if you want.

Thank you again for your time!

Samuel

### CODE REVIEW

File: gecode/int.hh [hunk 1]
```diff
@@ -5172,6 +5173,58 @@ namespace Gecode {
...
+   * currently supported for the following propagators:
+   *
+   *   - Gecode::Int::Distinct::Val
+   *   - Gecode::Int::Distinct::Bnd
+   *   - Gecode::Int::Distinct::Dom
+   *   - More to come...
+   *
+   * If the space does not contain any of the preceding propagators, this
+   * brancher won't be able to make any branching choices. For more details,
+   * please see the documentation in MPG, section ?.
```
Comment (SaGagnon) [inline gecode/int.hh]: I will make a pull request to add documentation in MPG.

File: gecode/int/branch.hh [hunk 2]
```diff
@@ -672,6 +672,123 @@ namespace Gecode { namespace Int { namespace Branch {
...
+     * all spaces.
+     */
+    class VarIdToPos : public SharedHandle {
+    protected:
+      class VarIdToPosO : public SharedHandle::Object {
+      public:
+        /// The map we want to share
+        std::unordered_map<unsigned int, unsigned int> _varIdToPos;
+      public:
+        /// Default constructur
```
Comment (SaGagnon) [inline gecode/int/branch.hh]: Typo... I will correct it.

File: gecode/int/branch.hh [hunk 3]
```diff
@@ -672,6 +672,123 @@ namespace Gecode { namespace Int { namespace Branch {
 
 #include <gecode/int/branch/view-values.hpp>
 
+#ifdef GECODE_HAS_CBS
```
Comment (SaGagnon) [inline gecode/int/branch.hh]: Is gecode/int/branch.hh a good place to put the definition of my brancher?

File: gecode/int/branch/cbs.hpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+              typename decltype(logProp)::size_type(),
+              typename decltype(logProp)::hasher(),
+              typename decltype(logProp)::key_equal(),
+              typename decltype(logProp)::allocator_type(home)) {
+    x.update(home,b.x);
+  }
+
+  template<class View>
+  bool CBSBrancher<View>::status(const Space& home) const {
+    auto& h = const_cast<Space&>(home);
```
Comment (SaGagnon) [inline gecode/int/branch/cbs.hpp]: There is no method for iterating in const propagators, so I needed to do this. In line 120, I tried to make explicit the fact that domainsizesum is a const method.

File: gecode/int.hh (hunk 1 above)
Comment (chschulte) [reply gecode/int.hh]: If you want, I can give you access to MPG. It is slightly complicated, though.

File: gecode/int/branch.hh (hunk 3 above)
Comment (chschulte) [reply gecode/int/branch.hh]: Yes, absolutely!

File: gecode/int.hh (hunk 1 above)
Comment (chschulte) [reply gecode/int.hh]: If you want, I can give you access to MPG. It is a little scary, though...

File: gecode/int/branch/cbs.hpp (hunk 4 above)
Comment (chschulte) [reply gecode/int/branch/cbs.hpp]: I just changed that in the kernel itself...

Comment (chschulte) [reply gecode/int/branch/cbs.hpp]: Which means you can remove the cast eventually

**Comment by chschulte** [comment]
If you think you are done, a pull request against develop would be nice. Please include the change I did to the kernel.

**Pull request merged**